text/0017-message/0017-message.md

```diff
@@ -0,0 +1,172 @@
+
+| Payload Name                              |   Type Value |
+| ---------------------------------------   | -----------  | 
+|  [Signed Transaction](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-signed-transaction-payload/0000-signed-transaction-payload.md)                       |     0        |
+|  [Milestone Draft](https://github.com/jakubcech/protocol-rfcs/blob/jakubcech-milestonepayload/text/0019-milestone-payload/0019-milestone-payload.md)                                |     1        |
+|  [Indexation Payload](#indexation-payload)  |     2        |
+
+### Indexation payload
+
+To make the Payload concept clear we will define the `indexation payload`. As the name suggests it allows to add an index to the encapsulating message, as well as some arbitrary data. Nodes will expose an API, that will enable to query messages by the index.
+Adding those capabilities may open nodes to DOS attack vectors:
+1. Proliferation of index keys that may blow up the node's DB
+2. Proliferation of messages associated with the same index
+
+Node implementations may provide weak guarantees regarding the completion of indexes to address the above scenarios. 
+
+Besides the index, the payload will also have a data field.
+  A message that has been attached to the tangle and approved by a milestone has useful properties: You can verify that the content of the data did not change, and you can ascertain the approximate time it was published by checking the approving milestone. If the payload will be incorporated under
+  the `signed transaction payload`, the content will be signed as well.
+
+
+The structure of the payload is simple:
+
+| Name             | Type          | Description               |
+| --------         | -----------   | -----------               |
+| Payload Type     | uint32        | Must be set to **2**      |
+| Index            | string     | The index key of the message |
+| Data             | ByteArray     | Data we are attaching    |
+
+Note that `index` and `data` may both have a length of 0.
+There are no validation rules for the payload. Message validation rules suffice here.
+
+### Serialization Example
+
+Below is a serialized valid message with the indexation payload. The index is the "SPAM" ASCII string and the message is the "Hello Iota"
+ASCII string. The [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) should have
+  14 trailing zeroes. Bytes are expressed as hexadecimal numbers. 
+
+[Version] **`01`** [Parent 1] `F532A53545103276B46876C473846D98648EE418468BCE76DF4868648DD73E5D` [Parent 2] `78D546B46AEC4557872139A48F66BC567687E8413578A14323548732358914A2` [Payload Length]
+*`0C`*[Payload Type]**`02`**[Index] *`04`*`9A52` [Data]*`0A`*`48656c6c6f20496f7461`[Nonce]`dc293f3333333333`
```

Comment:
`9A52` == `SPAM` ?

Reply:
Hmm...
Not sure how this happened :-|

Also the payload length is incorrect...
I will change the amount of trailing 0s to 10 for the example because it is easier to calculate

Reply:
Also, the index length and data length seem to be only one byte ?

Reply:
Before I fix, I am starting to wonder whether the 32 bytes we specified for the length is a good decision...
I understand it is convenient, but maybe it is just bloating up message size for no good reason?